python3Packages.formulaic: test failed

### Nixpkgs version

- Unstable (25.11)

### Describe the bug

related hydra log: https://hydra.nixos.org/build/302466817

test failure after version bump of numpy/scipy

### Steps to reproduce

nix build nixpkgs#python3Packages.formulaic

### Expected behaviour

build failure

### Screenshots

_No response_

### Relevant log output

```console
self = <tests.materializers.test_pandas.TestPandasMaterializer object at 0x7ffbed1da350>
data_with_nulls =      a  b     A     B  D
0  1.0  1     a     a  a
1  2.0  2  None     b  a
2  NaN  3     c  None  a
formula = 'A:a'
tests = (['Intercept', 'A[a]:a', 'A[b]:a', 'A[c]:a'], ['Intercept', 'A[a]:a', 'A[b]:a', 'A[c]:a'], ['Intercept', 'A[a]:a'], 1)
output = 'sparse'

    @pytest.mark.parametrize("formula,tests", PANDAS_TESTS.items())
    @pytest.mark.parametrize("output", ["pandas", "numpy", "sparse"])
    def test_na_handling(self, data_with_nulls, formula, tests, output):
        mm = PandasMaterializer(data_with_nulls).get_model_matrix(
            formula, output=output
        )
        if output == "pandas":
            assert isinstance(mm, pandas.DataFrame)
            assert list(mm.columns) == tests[2]
        elif output == "numpy":
            assert isinstance(mm, numpy.ndarray)
    
        else:
>           assert isinstance(mm, spsparse.csc_matrix)
E           AssertionError: assert False
E            +  where False = isinstance(<COOrdinate sparse matrix of dtype 'float64'\n with 2 stored elements and shape (1, 2)>, <class 'scipy.sparse._csc.csc_matrix'>)
E            +    where <class 'scipy.sparse._csc.csc_matrix'> = spsparse.csc_matrix

tests/materializers/test_pandas.py:215: AssertionError
=========================== short test summary info ============================
FAILED tests/materializers/test_pandas.py::TestPandasMaterializer::test_get_model_matrix_sparse[A:a-tests3] - AssertionError: assert False
FAILED tests/materializers/test_pandas.py::TestPandasMaterializer::test_get_model_matrix_sparse[A:B-tests4] - AssertionError: assert False
FAILED tests/materializers/test_pandas.py::TestPandasMaterializer::test_na_handling[sparse-A:a-tests3] - AssertionError: assert False
================== 3 failed, 488 passed, 11 skipped in 2.24s ===================
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.34, NixOS, 25.11 (Xantusia), 25.11.20250627.30e2e28`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.28.4`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/m3s9kr28wjhh52qkjrd9spnd4l6lprwf-nixpkgs/nixpkgs`

### Notify maintainers

cc @swflint 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+bug%22+-label%3A%226.topic%3A+darwin%22+-label%3A%226.topic%3A+nixos%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/formulaic/default.nix
+++ b/pkgs/development/python-modules/formulaic/default.nix
@@ -1,58 +1,68 @@
 {
   lib,
-  astor,
   buildPythonPackage,
+  pythonOlder,
   fetchFromGitHub,
   hatch-vcs,
   hatchling,
   interface-meta,
+  narwhals,
   numpy,
   pandas,
-  pytestCheckHook,
-  pythonOlder,
   scipy,
-  sympy,
   typing-extensions,
   wrapt,
+  pyarrow,
+  polars,
+  sympy,
+  pytestCheckHook,
 }:
 
 buildPythonPackage rec {
   pname = "formulaic";
-  version = "1.1.1";
+  version = "1.2.0";
   pyproject = true;
 
-  disabled = pythonOlder "3.7";
+  disabled = pythonOlder "3.9";
 
   src = fetchFromGitHub {
     owner = "matthewwardrop";
     repo = "formulaic";
     tag = "v${version}";
-    hash = "sha256-7vmnibL0PMZWL/unUQxN4GtLPSYKmGnhDNtE5GRDFHk=";
+    hash = "sha256-+MyMxtX3DCmluf+8dvo0uulc1AeKE6les1xiRGVWjPU=";
   };
 
   env.SETUPTOOLS_SCM_PRETEND_VERSION = version;
 
-  nativeBuildInputs = [
+  build-system = [
     hatchling
     hatch-vcs
   ];
 
-  propagatedBuildInputs = [
-    astor
+  dependencies = [
+    narwhals
     numpy
     pandas
     scipy
     wrapt
     typing-extensions
     interface-meta
-    sympy
   ];
 
-  nativeCheckInputs = [ pytestCheckHook ];
+  optional-dependencies = {
+    arrow = [ pyarrow ];
+    polars = [ polars ];
+    calculus = [ sympy ];
+  };
+
+  nativeCheckInputs =
+    [
+      pytestCheckHook
+    ]
+    ++ optional-dependencies.arrow
+    ++ optional-dependencies.calculus;
 
   pythonImportsCheck = [ "formulaic" ];
-
-  disabledTestPaths = [ "tests/transforms/test_poly.py" ];
 
   meta = with lib; {
     description = "High-performance implementation of Wilkinson formulas";

